### ISSUE TITLE
Kicking things off

### ISSUE THREAD

**Issue by rgbkrk** [nteract/models#1]
We've done some previous explorations for a model backed by `setIn`

Originally it started with the [synchronized object diff spec](https://github.com/rgbkrk/synchronized-object-diff-spec). We evenutally implemented a "hidden" version of this that uses a `setIn` based reducer under the hood:

https://github.com/nteract/nteract/blob/master/packages/core/src/reducers/comms.js#L42-L55

Example notebook: https://github.com/nteract/nteract/blob/master/applications/desktop/example-notebooks/model-debug.ipynb

----

This repository is going to be an exploration on a python module for binding the comms (two-way) with a python object for these. 

cc @mpacer @rmorshea @ivanov

**Comment by rmorshea** [comment]
@ivanov since the Paris widgets workshop in January I've been working on a package called [`spectate`](https://python-spectate.readthedocs.io/en/latest/) that could drive an event system for Python. I initially demoed `spectate` in the context of [traitlets](https://github.com/ipython/traitlets) and `ipywidgets` calling the resulting classes ["eventful" traits](https://github.com/ipython/traitlets/pull/466), however it seems pretty unlikely that the proposal will be merged any time soon.

Since then, I met with @rgbkrk and @mpacer who encouraged me to contribute here. Given the relatively fresh slate that we'll be working with, I've been rethinking that proposal, and I decided to use `spectate` to create [a foundation](https://github.com/rmorshea/spectate#the-bleeding-edge) for the MVC framework that we'll need going forward.

Looking forward to all of your feedback.

**Comment by rgbkrk** [comment]
We met in person and each shared our own demos and thinking. @rmorshea's `spectate` library, @gnestor's `target_name` + event proposal, @rgbkrk's `setIn` + data attribute proposal.

Since I'm now filling this out _after_ we met yesterday I'm going to miss some parts. Please feel free to edit this comment, comment below, or let me know what else you think.

## `spectate`

**Proposer: @rmorshea**

spectate provides a way to watch for changes and trigger a callback with old and new values:

```python
from spectate import mvc

d = mvc.Dict()

@mvc.view(d)
def printer(event):
    print(event)

d['a'] = 1
```

```python
{'key': 'a', 'old': Undefined, 'new': 1}
```

There's more available within spectate as well, and could be used as a way to create models.

Ryan also showed off how we could update children elements within a VDOM structure.

## `target_name` backed events

**Proposer: @gnestor**

https://github.com/nteract/vdom/pull/44

When an attribute that is an callback `onClick` is set to an object of the form:

```
{
  'target_name': 'some-target'
}
```

It automatically sends over the Synthetic Event from that callback on comm with `target_name`. Note: no `comm_id` is assumed.

## data paths, setIn

**Proposer: @rgbkrk**

Write some `data-` attributes to specify the 

* attribute
* model (comm id)
* path on the model 

```html
<h1 data-model-for-children="model:comm-1234;path:a">original children</h1>
```

Which will go along with the `setIn(path, value)` backed models like this:

```
models = {
  "comm-1234": {
    "a": "woohoo"
  }
}
```

When rendered, the model and path is looked up to replace `children` (because we used `data-model-for-children`):

```html
<h1 data-model-for-children="model:1234;path:a">woohoo</h1>
```

When this is set on an attribute that is modifiable by a user e.g. `value`

```html
<input data-model-for-value="model:1234;path:a" value="test" />
```

an `onChange` is created that will allow the frontend to send back:

```
{
  target_module: "reducers",
  target_name: "setIn",
  path: "a",
  value: (value from event.target.value)
}
```

## Dash components

We all seem to love the dash components and need to dig into them. 😄

**Comment by rmorshea** [comment]
@rgbkrk I'm pretty convinced that what I did with `spectate` isn't useful here. After seeing @gnestor's work I came to realize that `spectate` provides a much more general solution to data synchronization than is really necessary. `spectate` would enable us to sync *all* data between the frontend and backend, but we really only need to sync *certain* things.

**Comment by rgbkrk** [comment]
I totally agree @rmorshea

**Comment by gnestor** [comment]
My goal with this is to provide a _React-like_ experience in Python. vdom currently does this for rendering native DOM elements but doesn't support event handlers. https://github.com/nteract/vdom/pull/44 and https://github.com/jupyterlab/jupyter-renderers/pull/134 add event handler support to vdom and jupyter-renderers, respectively. This allows events to be handled in Python and in [my demo](https://mybinder.org/v2/gh/gnestor/jupyter-renderers/vdom-experimental?urlpath=lab/tree/notebooks/vdom-experimental-extension.ipynb) those Python callbacks are used to call `IPython.display.update_display` to update output on the front-end. 

After discussing in person, @rgbkrk suggested that we write up a spec of event types and their comm payload because a) we can't serialize DOM event entirely (for transport over comms) and b) smaller payloads are faster, and c) events contain a lot of irrelevant data. The React [SyntheticEvent docs](https://reactjs.org/docs/events.html#reference) have done a lot of this work for us already: they list all supported event types (e.g. `KeyboardEvent`), event names (e.g. `onKeyDown`), and event properties (e.g. `keyCode`). I can create a markdown doc that transfers most of this minus the event properties that are DOM nodes (e.g. `view`). 

Is this still worth doing?

**Comment by rgbkrk** [comment]
Yep! Creating a markdown doc of the event types we'll support and the keys would be mega fantastic. After that we can make json schema for it.

**Comment by rmorshea** [comment]
@gnestor @rgbkrk I'm trying to taking my `spectate` prototype to some state of completion through [Purly](https://github.com/rmorshea/purly). My goal has been to create a standalone library that will allow Python users with minimal JS/HTML skills, to control the DOM. I think Purly is somewhat tangential to the work here since I wanted to put Python in the drivers seat.

What's relevant here though is Purly's generalized [model server](https://github.com/rmorshea/purly/blob/master/purly/model/server.py) that facilitates data synchronization and bidirectional communication between clients. I've only just started to document things so I'd be happy to hop on a call if you all have questions. In leu of that though I've written up a [basic example](https://mybinder.org/v2/gh/rmorshea/purly/master?filepath=examples/notebooks/introduction.ipynb) that shows off what it can do.
